### ISSUE TITLE
Page in PDF not loading completely in Firefox, missing image in Chromium

### ISSUE THREAD

**Issue by pcworld** [mozilla/pdf.js#8076]
Link to PDF file: https://www.personalausweisportal.de/SharedDocs/Downloads/DE/Flyer-und-Broschueren/eID_Broschuere.pdf?__blob=publicationFile&v=1 (sha512sum 62e4a6a96257f219d1f1fc1c644eea86b17c056486a91a159e4cc0085347b3bc3c3d1a4e06eae204e1dce92f507afa5d55c21604584ffa852efb7fb303f9f846)

Configuration:
- Web browser and its version: Chromium 53.0.2785.143-0ubuntu0.14.04.1.1145, Firefox 51.0.1+build2-0ubuntu0.14.04.2
- Operating system and its version: Ubuntu 14.04
- PDF.js version: https://mozilla.github.io/pdf.js/web/viewer.html
- Is an extension: Issues happen in the Chromium pdf.js extension, Firefox's native pdf.js and the aforementioned online version of pdf.js

Steps to reproduce the problem:
1. Open the PDF in pdf.js and scroll to the last page.

What is the expected behavior?
The last page renders like in the following screenshot of the page in Evince:
![Last page in Evince](https://cloud.githubusercontent.com/assets/1506734/23024514/f8cffa64-f45a-11e6-8329-89f921ad7ad3.png)

What went wrong?
In Firefox: Page doesn't finish loading (loading animation doesn't disappear), page is only partially loaded:
![Last page in pdf.js in Firefox](https://cloud.githubusercontent.com/assets/1506734/23024575/315df3fe-f45b-11e6-847e-8111c26da84e.png)
In Chromium: Loading animation disappears, but the QR code is blank (white):
![Last page in pdf.js in Chromium](https://cloud.githubusercontent.com/assets/1506734/23024624/630f7a30-f45b-11e6-8b39-fd5281cfbcd5.png)

The PDF appears to have been created by "Adobe InDesign CS5.5 (7.5.3)".

**Comment by yurydelendik** [comment]
Hmm, that's QR code size is 16667x16667, it's like 8000dpi image.

**Comment by yurydelendik** [comment]
@pcworld can you check if you can reduce size of the image to acceptable dpi --perhaps 300dpi. I'm sure other PDF readers and network providers will say thanks :)

**Comment by pcworld** [comment]
@yurydelendik I am not the creator of this PDF.
While the resolution of the image is indeed ridiculous, Evince somehow renders it in a few seconds only.

**Comment by yurydelendik** [comment]
Yeah, there is probably optimization to render such images at lower resolution we are missing. I'm marking the issue with performance tag.

**Comment by Rob--W** [comment]
I have extracted the relevant page: [eID_Broschuere-page16.pdf](https://github.com/mozilla/pdf.js/files/1135236/eID_Broschuere-page16.pdf)

The culprit is the image mask of size 16667x16667. We should scale down the image.

```
16 0 obj
<< /BitsPerComponent 1 /DecodeParms << /Columns 16667 /K -1 >> /Filter /CCITTFaxDecode /Height 16667 /ImageMask true /Subtype /Image /Type /XObject /Width 16667 /Length 38972 >>
stream
```

Implementation details:

- The image is generated at https://github.com/mozilla/pdf.js/blob/85161fa6dc81c3df921b27a378fbaf82491aa9ee/src/core/evaluator.js#L356-L385
- These functions are responsible for painting the mask on the canvas: https://github.com/mozilla/pdf.js/blob/85161fa6dc81c3df921b27a378fbaf82491aa9ee/src/display/canvas.js#L1942-L1980 and https://github.com/mozilla/pdf.js/blob/85161fa6dc81c3df921b27a378fbaf82491aa9ee/src/display/canvas.js#L2079-L2148
- In the SVG backend (e.g. used by examples/node/pdf2svg.js), the image is rendered at https://github.com/mozilla/pdf.js/blob/85161fa6dc81c3df921b27a378fbaf82491aa9ee/src/display/svg.js#L1079-L1131

To scale down the image, it is probably best to detect large images, slice the image in pieces, transforming (=scaling) all those images individually (on a canvas) and then painting all images together. An alternative approach is to manually transform the image, i.e. interpreting the pixels of the image yourself and interpolate the pixel values while scaling. The advantage of the latter is that its runtime performance is likely better for arbitrarily large images, and that the logic can be shared by our canvas and SVG backends (since this would then more be a math problem than a rendering task).

Debugging tips:

If you are going to debug this issue with a debugger, consider adding `#disableWorker=true` to the URL. Otherwise you have to account for the fact that the logic of `src/core` runs in a Web Worker, while the canvas logic runs on the main thread.

**Comment by apoorv-mishra** [comment]
Hi, I would like to work on this issue. But, I have got a couple of questions regarding @Rob--W 's comment above -
> To scale down the image, it is probably best to detect large images...
1. What's the benchmark for _large_ images? It would be helpful if you can specify the size (in `w` x `h`) which may be labelled as _large_.
> An alternative approach is to manually transform the image, i.e. interpreting the pixels of the image yourself and interpolate the pixel values while scaling.
2. What do you mean by _interpreting the pixels_ and _interpolate the pixel values_? Are you referring to Bilinear Interpolation? Would the implementation be similar to https://github.com/mozilla/pdf.js/blob/master/src/core/colorspace.js#L34-L56?

Also, if possible, please guide me with providing appropriate resources which I can study from and go about implementing the solution.

**Comment by Rob--W** [comment]
Hi @apoorv-mishra 

> > To scale down the image, it is probably best to detect large images...
> 1. What's the benchmark for _large_ images? It would be helpful if you can specify the resolution(in `w` x `h`) which may be labelled as _large_.

I don't have a specific value in mind, but I was thinking of big images whose width/height are significantly larger than the actually painted image, to the extend that the native image decoder would be unable to handle it, or that it would require an excessive amount of memory). You can start with a hard-coded value (that you find experimentally by trying to paint images of that size, and/or by looking at other parts of PDF.js where a maximum image/canvas size is enforced).
If needed, we can make the logic more complex later (e.g. deciding whether an image is too large based on the actual size of the painted image in the rendered PDF). But for now, let's keep it simple and choose a reasonable hard-coded threshold.

> > An alternative approach is to manually transform the image, i.e. interpreting the pixels of the image yourself and interpolate the pixel values while scaling.
> 2. What do you mean by _interpreting the pixels_ and _interpolate the pixel values_? Are you referring to Bilinear Interpolation? Would the implementation be similar to https://github.com/mozilla/pdf.js/blob/master/src/core/colorspace.js#L34-L56?

By interpreting, I mean code that takes the pixel data and does something with it.
By interpolating, I mean interpolation in the mathematical sense. That is, take a (large) group of pixels, calculate one pixel value that approximates the appearance of the original set of pixels. Bilinear interpolation is one of the possible ways to do it, you need to investigate the available options and see which one results in a drawing that is the best approximation of the original image. Search for "browser image resize algorithm" in your favorite search engine to learn more about how browsers scale images
The `resizeRgbImage` function that you linked indeed looks like a good start.

Tip: When you link to code online, link to a specific commit instead of a branch (like "master"), because the code in the master branch can change and the line numbers can become different). On Github you can get a specific commit by pressing the "y" key. So https://github.com/mozilla/pdf.js/blob/master/src/core/colorspace.js#L34-L56 becomes https://github.com/mozilla/pdf.js/blob/5b5781b45d234666241bf3354c0d390315c31d1a/src/core/colorspace.js#L34-L56

> Also, if possible, please guide me with providing appropriate resources which I can study from and go about implementing the solution.

The code that I linked in my previous comments is a good start. At that point, the problem space has already been reduced from "an image in a PDF file" to "an image to be displayed". You don't need specialized knowledge of PDF to implement this.

If you want to go with the first method, you need to know how to work with the canvas API - https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API. The PDF.js code base already has several eexamples.
To implement the alternative approach, you need to know what the bytes of the image data mean. I don't know this from the top of my head, but if you step through the code with a debugger you can probably see some useful information.

**Comment by Snuffleupagus** [comment]
@calixteman Will this issue also be fixed by PR #16077?
The link above seems to be broken, but the document is available (as `issue8076.pdf`) in the PDF archive I shared a while back.

**Comment by calixteman** [comment]
Yep it's one of the files I used (and the pdf can be found in https://github.com/mozilla/pdf.js/issues/8076#issuecomment-314078517).
I should add it in the test suite since it takes a specific path (it's a mask):
https://github.com/mozilla/pdf.js/blob/d7e4be9cdbe37d4d4d9ada34208820470bcd14ed/src/core/image.js#L374
